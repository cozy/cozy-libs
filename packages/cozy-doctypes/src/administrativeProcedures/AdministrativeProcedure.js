const get = require('lodash/get')
const flatten = require('lodash/flatten')

const Contact = require('../contacts/Contact')
const Document = require('../Document')

class AdministrativeProcedure extends Document {
  /**
   * Returns personal data for the contact
   *
   * @param {Contact} contact - A contact
   * @param {Array} fields - The list of fields to retrieve
   * @return {Object} - the personal data
   **/
  static getPersonalData(contact, fields) {
    const mapping = {
      firstname: {
        path: 'name.givenName'
      },
      lastname: {
        path: 'name.familyName'
      },
      address: {
        getter: Contact.getPrimaryAddress
      },
      email: {
        getter: Contact.getPrimaryEmail
      },
      phone: {
        getter: Contact.getPrimaryPhone
      }
    }
    let personalData = {}
    fields.forEach(field => {
      const contactField = get(mapping, field, field)
      let value
      if (contactField.getter) {
        value = contactField.getter(contact)
      } else {
        const path = get(contactField, 'path', field)
        value = get(contact, path)
      }

      if (value !== undefined) {
        personalData[field] = value
      }
    })

    return personalData
  }
  /**
   * Method to generate a query based on a few rules given by the template
   * @param {Object} docRules
   * @param {Object} docRules.rules
   * @param {int} docRules.count
   */
  static async getFilesByRules(docRules) {
    const { rules, count } = docRules
    const cozyRules = {
      trashed: false,
      type: 'file',
      ...rules
    }
    //Create an index in order to query and sort
    await this.cozyClient
      .collection('io.cozy.files')
      .createIndex(['metadata.datetime', 'metadata.classification'])
    //Use the index
    const files = await this.cozyClient
      .collection('io.cozy.files')
      .find(cozyRules, {
        indexedFields: ['metadata.datetime', 'metadata.classification'],
        sort: [
          {
            'metadata.datetime': 'desc'
          },
          {
            'metadata.classification': 'desc'
          }
        ],
        limit: count ? count : 1
      })

    return files
  }

  /**
   * Returns a io.cozy.procedures.administratives object
   *
   * @param {object} data - The data we need for this type of procedure
   * @param {ProcedureTemplate} template - The procedure's template
   * @return {AdministrativeProcedure} the administrative procedure
   */
  static create(data, template) {
    const { documentsData, personalData, procedureData } = data
    const files = Object.keys(documentsData).map(identifier => {
      return documentsData[identifier].files.map(file => {
        //TODO Remove this check. it has to be done before
        if (file)
          return {
            _id: file.id,
            _type: 'io.cozy.files',
            templateDocumentId: identifier
          }
      })
    })
    return {
      personalData,
      procedureData,
      submissionDate: new Date(),
      templateId: template.type,
      templateVersion: template.version,
      relationships: {
        files: {
          data: flatten(files)
        }
      }
    }
  }

  /**
   * Returns json that represents the administative procedure
   *
   * @param {AdministrativeProcedure}
   * @return {string} - the json that represents this procedure
   *
   */
  static createJson(administrativeProcedure) {
    return JSON.stringify(administrativeProcedure)
  }
}

AdministrativeProcedure.doctype = 'io.cozy.procedures.administratives'

module.exports = AdministrativeProcedure
