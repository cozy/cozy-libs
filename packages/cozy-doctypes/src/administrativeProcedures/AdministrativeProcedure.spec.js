import MockDate from 'mockdate'

import AdministrativeProcedure from './AdministrativeProcedure'

const MOCKED_DATE = '2018-01-01T12:00:00.210Z'

beforeAll(() => {
  MockDate.set(MOCKED_DATE)
})

afterAll(() => {
  jest.restoreAllMocks()
  MockDate.reset()
})

describe('AdministrativeProcedure model', () => {
  describe('getPersonalData', () => {
    it('should return the personal data for the given contact', () => {
      const dany = {
        fullname: '',
        name: {
          givenName: 'Daenerys',
          familyName: 'Targaryen'
        },
        phone: [
          {
            number: '+33 (2)0 90 00 54 04',
            primary: true
          },
          {
            number: '+33 6 77 11 22 33',
            primary: false
          }
        ],
        address: [
          {
            formattedAddress: '94 Hinton Road 05034 Fresno, Singapore',
            type: 'Home',
            primary: true
          },
          {
            street: '426 Runolfsson Knolls',
            city: 'Port Easter',
            country: 'Cocos (Keeling) Islands',
            postcode: '84573',
            type: 'Work'
          }
        ],
        email: [
          {
            address: 'dany@dragons.com',
            type: 'personal',
            primary: false
          },
          {
            address: 'daenerys@dragonstone.westeros',
            primary: true
          }
        ]
      }
      const fields = [
        'lastname',
        'firstname',
        'salary',
        'address',
        'phone',
        'email'
      ]
      const expected = {
        lastname: 'Targaryen',
        firstname: 'Daenerys',
        address: '94 Hinton Road 05034 Fresno, Singapore',
        phone: '+33 (2)0 90 00 54 04',
        email: 'daenerys@dragonstone.westeros'
      }
      const result = AdministrativeProcedure.getPersonalData(dany, fields)
      expect(result).toEqual(expected)
    })
  })

  describe('create', () => {
    it('should return an io.cozy.procedures.administratives object', () => {
      const data = {
        procedureData: {
          amount: 5000,
          duration: 36
        },
        personalData: {
          firstname: 'John',
          lastname: 'Doe',
          email: 'john.doe@me.com'
        },
        documents: []
      }
      const template = {
        type: 'credit-application',
        version: 1
      }
      const result = AdministrativeProcedure.create(data, template)
      expect(result).toEqual({
        procedureData: {
          amount: 5000,
          duration: 36
        },
        personalData: {
          firstname: 'John',
          lastname: 'Doe',
          email: 'john.doe@me.com'
        },
        documents: [],
        submissionDate: new Date(MOCKED_DATE),
        templateId: 'credit-application',
        templateVersion: 1
      })
    })
  })

  describe('createJson', () => {
    it('should return the json for this procedure', () => {
      const procedure = {
        procedureData: {
          amount: 5000,
          duration: 36
        },
        personalData: {
          firstname: 'John',
          lastname: 'Doe',
          email: 'john.doe@me.com'
        },
        documents: [],
        submissionDate: new Date(MOCKED_DATE),
        templateId: 'credit-application',
        templateVersion: 1
      }
      const result = AdministrativeProcedure.createJson(procedure)
      // turn the json back to an object for snapshot readability
      expect(JSON.parse(result)).toMatchSnapshot()
    })
  })
})
