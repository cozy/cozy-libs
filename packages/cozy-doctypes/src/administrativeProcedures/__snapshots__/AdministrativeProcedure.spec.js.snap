// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdministrativeProcedure model createJson should return the json for this procedure 1`] = `
Object {
  "documents": Array [],
  "personalData": Object {
    "email": "john.doe@me.com",
    "firstname": "John",
    "lastname": "Doe",
  },
  "procedureData": Object {
    "amount": 5000,
    "duration": 36,
  },
  "submissionDate": "2018-01-01T12:00:00.210Z",
  "templateId": "credit-application",
  "templateVersion": 1,
}
`;
